control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -154,11 +189,52 @@ impl ResourceLifecycleWithLifetime for OperationGuardArc<VolumeSnapshot> {
     async fn destroy(
         &mut self,
         registry: &Registry,
-        _request: &Self::Destroy<'_>,
+        request: &Self::Destroy,
     ) -> Result<(), SvcError> {
-        self.start_destroy_by(registry, &()).await?;
-        let result = Ok(());
-        self.complete_destroy(result, registry).await
+        self.start_destroy(registry).await?;
+        // Get the volume snapshot persisted info.
+        let volume_snapshot = request.volume_snapshot.lock().clone();
+
+        // Create a map to track the transactions upon deletion.
+        let mut updated_transactions: HashMap<SnapshotTxId, Vec<ReplicaSnapshot>> = HashMap::new();
+
+        // Iterate over current transactions and destroy each replica snapshot in a transaction,
+        // store the transactions and replica snapshots whose deletion failed.
+        for (txn_id, snap_list) in volume_snapshot.metadata().transactions() {
+            for snap in snap_list {
+                if let Err(err) =
+                    OperationGuardArc::<VolumeSnapshot>::destroy_replica_snapshot(snap, registry)
+                        .await
+                {
+                    if !matches!(err.tonic_code(), tonic::Code::NotFound) {
+                        let mut snap = snap.clone();
+                        // If it was an error mark the spec as deleting and push it in a map.
+                        snap.set_status_deleting();
+                        updated_transactions
+                            .entry(txn_id.clone())
+                            .or_insert_with(Vec::new)
+                            .push(snap);
+                    }
+                }
+            }
+        }
+
+        if updated_transactions.is_empty() {
+            // If there we no errors remove the persisted spec and complete destroy.
+            self.complete_destroy(Ok(()), registry).await?;
```

Comment:
Should we also set the transactions to empty in case persisting to pastor fails? Though that may not be needed, can't remember exactly how that is handled without looking at the code.

Reply:
I think the garbage collector will anyhow retry as we already started destroy. So it's probably not needed?

Reply:
It will retry but it's pointless to retry things which were already destroyed? Worst case let's say we did destroy it and the node is gone for some reason, the destroy will never succeed.

Reply:
Right. I will set it to empty.

Reply:
Non related: this reminds me we really need to handle "node gone" and "pool gone" scenarios, atm resources will get stuck and unable to be deleted in such scenarios :(